src/hashtable.c

```diff
@@ -0,0 +1,163 @@
+/*= -*- c-basic-offset: 4; indent-tabs-mode: nil; -*-
+ *
+ * hashtable.c -- a generic hashtable implementation.
+ *
+ * Copyright (C) 2016 by Donovan Baarda <abo@minkirri.apana.org.au>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include "hashtable.h"
+
+/* Open addressing works best if it can take advantage of memory caches using
+ * locality for probes of adjacent buckets on collisions. So we pack the keys
+ * tightly together in their own key table and avoid referencing the element
+ * table and elements as much as possible. Key value zero is reserved as a
+ * marker for an empty bucket to avoid checking for NULL in the element table.
+ * If we do get a hash value of zero, we -1 to wrap it around to 0xffff. */
+
+/* Use max 0.8 load factor to avoid bad open addressing performance. */
+#define HASHTABLE_LOADFACTOR_NUM 8
+#define HASHTABLE_LOADFACTOR_DEN 10
+
+hashtable_t *hashtable_new(int size, hash_f hash, cmp_f cmp)
+{
+    hashtable_t *t;
+    int size2;
+
+    /* Adjust requested size to account for max load factor. */
+    size = 1 + size * HASHTABLE_LOADFACTOR_DEN / HASHTABLE_LOADFACTOR_NUM;
+    /* Use next power of 2 larger than the requested size. */
+    for (size2 = 1; size2 < size; size2 <<= 1) ;
+    if (!(t = calloc(1, sizeof(hashtable_t)+ size2 * sizeof(unsigned))))
+        return NULL;
+    if (!(t->etable = calloc(size2, sizeof(void *)))) {
+        free(t);
+        return NULL;
+    }
+    t->size = size2;
+    t->count = 0;
+    t->hash = hash;
+    t->cmp = cmp;
+#ifndef HASHTABLE_NSTATS
+    t->find_count = t->match_count = t->keycmp_count = t->entrycmp_count = 0;
+#endif
+    return t;
+}
+
+void hashtable_free(hashtable_t *t)
+{
+    if (t) {
+        free(t->etable);
+        free(t);
+    }
+}
+
+/* MurmurHash3 finalization mix function. */
+static inline unsigned mix32(unsigned int h)
+{
+    h ^= h >> 16;
+    h *= 0x85ebca6b;
+    h ^= h >> 13;
+    h *= 0xc2b2ae35;
+    h ^= h >> 16;
+    return h;
+}
+
+/* Get hash key, reserving zero as an empty bucket marker. */
+static inline unsigned get_key(const hashtable_t *t, const void *e)
+{
+    unsigned k = t->hash(e);
```

Comment:
I did a lot of dicking around with this and tried a heap of different things, some of which are visible in the commit history, and some of which are not. I tried a variety of things to avoid the function de-ref overheads and in the end was surprised at how little difference it made, though maybe that was early in the development before I'd optimized as much for the memory cache.

Trying it now with the following diff applied;

--- a/src/hashtable.c
+++ b/src/hashtable.c
@@ -81,7 +81,7 @@ static inline unsigned mix32(unsigned int h)
 /* Get hash key, reserving zero as an empty bucket marker. */
 static inline unsigned get_key(const hashtable_t *t, const void *e)
 {
-    unsigned k = t->hash(e);
+    unsigned k = *(unsigned *)e;
     return k ? k : -1;
 }

which removes the hash-func de-refencing and just assumes that the first field in the entry struct is the unsigned hash value (it is) we get the following performances;

### 512K/0.5 with no hash func

librsync: match statistics: signature[268697600 searches, 262144 (0.098%) matches, 313791392 (1.168x) weak sum compares, 518749 (0.193%) strong sum compares, 518434 (0.193%) strong sum calcs]
librsync: delta statistics: literal[8192 cmds, 268435456 bytes, 24576 cmdbytes] copy[2 cmds, 268435456 bytes, 0 false, 18 cmdbytes] speed[536.9 MB (8.3 MB/s) in, 268.5 MB (4.1 MB/s) out, 65 sec]

#### 716K/0.7

librsync: match statistics: signature[375756800 searches, 366592 (0.098%) matches, 1053222017 (2.803x) weak sum compares, 867330 (0.231%) strong sum compares, 866681 (0.231%) strong sum calcs]
librsync: delta statistics: literal[11456 cmds, 375390208 bytes, 34368 cmdbytes] copy[2 cmds, 375390208 bytes, 0 false, 18 cmdbytes] speed[750.8 MB (7.1 MB/s) in, 375.4 MB (3.5 MB/s) out, 106 sec]

Comparing this to the equivalent runs above, the execution time dropped for 512K/0.5 from 72secs to 65secs, and for 716K/0.7 from 108sec to 106sec. For low load factors it does make a small but measurable difference, but for higher load factors it's in the noise, and the execution time is completely dominated by the memory cache costs of doing on average 2.8 hash value comparisons per search.

I originally wanted to make the hashtable API as clean and generic as possible, which is why I went with the indirect function. I also at one point tried a different API where you provided the hash value to hashtable_add() and hashtable_find() and did the hashing externally. In the end I felt like the performance benefits were too small to bother.